libraries/tock-cells/src/map_cell.rs

```diff
@@ -80,27 +131,67 @@ impl<T> MapCell<T> {
 
     /// Puts a value into the `MapCell`.
     pub fn put(&self, val: T) {
-        self.occupied.set(true);
-        unsafe {
-            ptr::write(self.val.get(), MaybeUninit::<T>::new(val));
-        }
+        // This will ensure the value as dropped
+        self.replace(val);
     }
 
     /// Replaces the contents of the `MapCell` with `val`. If the cell was not
     /// empty, the previous value is returned, otherwise `None` is returned.
     pub fn replace(&self, val: T) -> Option<T> {
-        if self.is_none() {
-            self.put(val);
-            None
-        } else {
-            unsafe {
+        match self.occupied.get() {
+            Uninit => {
+                unsafe {
+                    // Safety: Because we are in the Uninit state, we are not writing over anything
+                    // that needs to be dropped
+                    ptr::write(self.val.get(), MaybeUninit::<T>::new(val));
+                }
+                None
+            }
+            Init => unsafe {
+                // Safety: Because we are not in the InitBorrowed state, nothing is currently also
+                // referencing this value
                 let result: MaybeUninit<T> = ptr::replace(self.val.get(), MaybeUninit::new(val));
                 // `result` is _initialized_ and now `self.val` is now a new uninitialized value
                 Some(result.assume_init())
+            },
+            InitBorrowed => {
+                // I know the point of MapCell was possibly also to reduce the source of panics,
+                // but this always unsafe. The description says it returns the value in the MapCell,
+                // so defaulting to just returning the argument might lead to unintended behavior.
+                // I say that any (old) code that tries to replace a borrowed value just fault,
+                // it is likely that it was broken anyway.
```

Comment:
Before we merge, we should update this comment to something like

```
// Trying to replace a borrowed value is invalid use of the API
```

And we should also update the doc comment on this method to indicate as such.